### PR TITLE
Tag and release D2 converted sources

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -12,12 +12,24 @@ workflows:
...
+        - d2tag:
+            requires:
+            - test-1.081-devel
+            - test-1.081-prod
+            - test-2.070-devel
+            - test-2.070-prod
+            - test-2.071-devel
+            - test-2.071-prod
+            - test-2.077-upstream-devel
+            - test-2.077-upstream-prod
```
Comment (leandro-lucarella-sociomantic) [inline .circleci/config.yml]: ouch, this is really inconvenient, there is a lot of repetition in here already with the compiler versions :unamused:

Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: I was trying to use yaml anchors for this too but looks like there is not way to append elements to anchored array, only to an object. So you can do this:

```yaml
list: &list
- one
- two

array: *list
```

but not this:

```yaml
list: &list
- one
- two

array:
<<: *list
- three
```

This really sucks but seems to be inherent problem with YAML based configuration (i.e. we will have the same problem in Travis too if it ever gets to the point of specifying step dependencies). Can't thing of any workaround other than generating config file by script.

This is one case where Jenkins Groovy-based pipeline approach really shines as superior.

Comment (leandro-lucarella-sociomantic) [reply .circleci/config.yml]: > i.e. we will have the same problem in Travis too

Not really, not in this case, as in Travis we group all the jobs in one step and we declare dependencies (well, order really, not dependencies) between stages, not individual jobs. Maybe you can do something like that?

Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: https://discuss.circleci.com/t/better-way-to-define-workflow-stages/19376 though I don't have high hopes (nothing in docs hints about it).

Comment (leandro-lucarella-sociomantic) [reply .circleci/config.yml]: This really sucks. It's a lot of boilerplate and very easy to miss some place when updating.

Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: It will also scale terribly if we ever want to add more jobs to second stage - you will need to list all requirements per _each_ second stage job!

Comment (leandro-lucarella-sociomantic) [reply .circleci/config.yml]: :man_facepalming:

Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: Got some suggestions from CircleCI staff, will try the tomorrow

**Comment by mihails-strasuns-sociomantic** [comment]
Updated, test tag in my fork - https://circleci.com/workflow-run/260c481e-fc69-4436-b689-68b0511a4716

**Comment by mihails-strasuns-sociomantic** [comment]
Another big CircleCI failure - it doesn't screen secret env vars from logs, you have to do it manually. Tweaking the script..

**Comment by mihails-strasuns-sociomantic** [comment]
Fixed, updated workflow - https://circleci.com/workflow-run/5604fde1-ce51-4427-95a7-686c153b77a7

File: .circleci/config.yml [hunk 2]
```diff
@@ -12,11 +12,25 @@ workflows:
...
+                tags:
+                    ignore: /.*\+d2$/
+                branches:
+                    ignore: /.*/
 
 templates:
     test: &test-template
         docker:
-        - image: sociomantictsunami/dlang:xenial-v3.2.0-rc.1
+        - image: sociomantictsunami/dlang:xenial-v3.2.0
```
Comment (leandro-lucarella-sociomantic) [inline .circleci/config.yml]: Why not just `v3` to get automatic updates? Minor versions should be reasonably reliable, but we could also start tagging `v3.2` too if you prefer to only get patch-updates.

File: .circleci/d2tag.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
+set -e
+
+# Performs a GitHub API call
```
Comment (leandro-lucarella-sociomantic) [inline .circleci/d2tag.sh:3]: Why not beaver? As a first hack you could just assign the `TRAVIS_*` variables from the contents of `CIRCLE_` variables. Then if this works some auto-detection can be added to beaver (hopefully temporarily until we move away from Travis, if CircleCI proves to be a better alternative).

File: .circleci/config.yml (hunk 1 above)
Comment (leandro-lucarella-sociomantic) [reply .circleci/config.yml]: No luck to avoid the repetition?

File: .circleci/config.yml [hunk 4]
```diff
@@ -12,11 +12,25 @@ workflows:
...
         - test-2.077-upstream-prod
+        - d2tag:
+            context: org-global
+            requires:
+            - test-2.070-devel
+            - test-2.070-prod
+            - test-2.071-devel
+            - test-2.071-prod
+            - test-2.077-upstream-devel
+            - test-2.077-upstream-prod
```
Comment (leandro-lucarella-sociomantic) [inline .circleci/config.yml]: I would depend also on D1, do we really want to release a D2 tag if the D1 tag is broken? This would also mean a bit less of special casing.

File: .circleci/config.yml (hunk 2 above)
Comment (leandro-lucarella-sociomantic) [inline .circleci/config.yml]: Ideally this should be changed in a separate commit (the changing of the tag, I'm not sure GitHub will display this comment in the right place :).

File: .circleci/config.yml [hunk 5]
```diff
@@ -48,6 +21,37 @@ templates:
         - run:
             name: Upload code coverage report
             command: bash <(curl -s https://codecov.io/bash)
+    job: &job-template
```
Comment (leandro-lucarella-sociomantic) [inline .circleci/config.yml]: Not important, but this labelling of templates confused me at first, I was looking for places where `job` was used. In the Travis jobs I used just a list, IMHO is a bit less noisy:
```yaml
templates:
    - &test-template
        docker: ...
    - &job-template
        docker: ...
```

File: .circleci/config.yml (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: Nope. I can use yaml anchors to paste whole array of dependencies but it will only make difference when we have more than one job at stage 2.

File: .circleci/config.yml (hunk 5 above)
Comment (mihails-strasuns-sociomantic) [reply .circleci/config.yml]: Nice, it did not come to my mind this is possible

File: .circleci/d2tag.sh (hunk 3 above)
Comment (mihails-strasuns-sociomantic) [reply .circleci/d2tag.sh:3]: I have removed beaver submodule completely for now and copying necessary utils from it on per need basis. Reason is that beaver is currently not developed with CircleCI specifics in mind and I don't want to risk something accidentally getting exposes because of seemingly innocent change upstream. When it will be time to decide if we keep or drop CircleCI here, we shall modify turtle/beaver accordingly, but until that point it is more convenient this way.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated

File: .circleci/d2tag.sh (hunk 3 above)
Comment (leandro-lucarella-sociomantic) [reply .circleci/d2tag.sh:3]: Sounds fair 👍

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
OK, I'm not happy about the excessive verboseness of the dependencies, but let's move forward with this!
